homeassistant/components/sensor/gitlab_ci.py

```diff
@@ -0,0 +1,187 @@
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            ATTR_BUILD_STATUS: self._status,
+            ATTR_BUILD_STARTED: self._started_at,
+            ATTR_BUILD_FINISHED: self._finished_at,
+            ATTR_BUILD_DURATION: self._duration,
+            ATTR_BUILD_COMMIT_ID: self._commit_id,
+            ATTR_BUILD_COMMIT_DATE: self._commit_date,
+            ATTR_BUILD_ID: self._build_id,
+            ATTR_BUILD_BRANCH: self._branch
+        }
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        if self._status == 'success':
+            return ICON_HAPPY
+        if self._status == 'failed':
+            return ICON_SAD
+        return ICON_OTHER
+
+    def update(self):
+        """Collect updated data from GitLab API."""
+        self._gitlab_data.update()
+
+        self._status = self._gitlab_data.status
+        self._started_at = self._gitlab_data.started_at
+        self._finished_at = self._gitlab_data.finished_at
+        self._duration = self._gitlab_data.duration
+        self._commit_id = self._gitlab_data.commit_id
+        self._commit_date = self._gitlab_data.commit_date
+        self._build_id = self._gitlab_data.build_id
+        self._branch = self._gitlab_data.branch
+        self._state = self._gitlab_data.status
+        self._available = self._gitlab_data.available
+
+
+class GitLabData():
+    """GitLab Data object."""
+
+    def __init__(self, gitlab_id, priv_token, interval, url):
+        """Fetch data from GitLab API for most recent CI job."""
+        self._gitlab_id = gitlab_id
+        self._priv_token = priv_token
+        self._interval = interval
+        self._url = url
+        self.update = Throttle(interval)(self._update)
+
+        self._response = None
+        self._response_json = None
+        self.available = False
+        self.status = STATE_UNKNOWN
+        self.started_at = STATE_UNKNOWN
+        self.finished_at = STATE_UNKNOWN
+        self.duration = STATE_UNKNOWN
+        self.commit_id = STATE_UNKNOWN
+        self.commit_date = STATE_UNKNOWN
+        self.build_id = STATE_UNKNOWN
+        self.branch = STATE_UNKNOWN
+        self.state = STATE_UNKNOWN
+
+    def _update(self):
+        _logger = logging.getLogger(__name__)
+        _logger.debug(self._interval)
+        import gitlab
+        import requests
+        try:
+            _gitlab = gitlab.Gitlab(
+                self._url, private_token=self._priv_token)
+            _gitlab.auth()
+            _projects = _gitlab.projects.get(self._gitlab_id)
+            _last_pipeline = _projects.pipelines.list()[0]
+            _last_job = _last_pipeline.jobs.list()[0]
```

Comment:
Do you really need to fetch _all_ jobs to get the first item?

Reply:
Changed call to request just first page of results, and only 1 result per page, thus requesting exactly what is needed and nothing more.